### ISSUE TITLE
How does Firejail compared to Sandboxie

### ISSUE THREAD

**Issue by Raj2032** [netblue30/firejail#2073]
There is a Windows sandboxing application called 'Sandboxie' https://www.sandboxie.com. How does it differ to Firejail (other than limiting you that you can only sandbox one application at a time with the free version for Sandboxie)?

**Comment by chiraag-nataraj** [comment]
I think one of the main differences is that `firejail` uses established technologies implemented at the kernel level, while `sandboxie` seems to implement stuff on its own. Also, `firejail` lets you isolate things like network stacks, which don't seem to be under the purview of `sandboxie`. Further, you can set resource controls (how much RAM, CPU, etc a sandbox can use) in `firejail`. Also, `firejail` lets you set up user namespaces (which users are visible in the sandbox), which, again, don't seem to even register in `sandboxie`.

TL;DR: Because `firejail` uses existing kernel technologies, it is (as far as I can tell) far more capable and exhaustive than `sandboxie`.

**Comment by chiraag-nataraj** [comment]
Here's a detailed comparison (`--these` are command-line arguments you can use for `firejail`):
**Appearance Settings**
Not currently in firejail (see #1983)

**Recovery Settings**
Usually done with `--whitelist` (for folders where you want to allow the program to see and write to) and `--get` to get files from other places within the sandbox. There are no exact analogues of these settings however (since `firejail` goes to sleep after launching the program).

**Delete Settings**
No exact analogue in `firejail` — all files created by the sandbox are deleted by default with the exception of stuff in `--whitelist`ed directories. The command used to delete the sandbox is not configurable in `firejail`.

**Program Start Settings**
No analogue to **Forced Folders** settings. **Forced Programs** is similar to `sudo firecfg`, which sets it up so that all programs shipped with `firejail` profiles are automatically sandboxed by default. You can do this manually (i.e. for only _certain_ programs) by doing `sudo ln -s /usr/bin/firejail /usr/local/bin/<program name>`.

**Program Stop Settings**
No analogue to these settings, but this has been an issue in the past (see #725).

**File Migration Settings**
No analogue to this setting. All files which are whitelisted will be copied over.

**Restrictions Settings**
**Internet Access** <-> `--net=none`, but `--net` is more powerful than that.
**Start/Run Access** <-> No clear analogue, but we have `--private-bin`, which functions in a similar way.
**Drop Rights** <-> `--noroot`, `--nonewprivs`
Other than these, there are _far_ more options — `--private-etc`, `--private-lib`, `--private-dev`, `--private-srv`, `--caps{,.drop,.keep}`, `--seccomp{,.block-secondary,.drop,.keep}`, and so on. Check the manual page (`man firejail`) for more info.

**Resource Access Settings**
**File Access > Direct Access & File Access > Full Access** <-> `--whitelist` (the distinction doesn't apply in `firejail`).
**File Access > Read-Only Access** <-> `--read-only`.
**File Access > Write-Only Access** <-> No exact analogue, but can use judicious `--whitelist` commands to achieve a similar effect.
**File Access > Blocked Access** <-> `--blacklist`.
**Registry Access** doesn't apply.
**IPC Access** <-> `--ipc-namespace`
**Window Access** <-> Can kind of achieve a similar effect using `--x11`.
**COM Access** doesn't apply.

**Applications Settings**
This is basically a way to edit the profiles that are shipped with `sandboxie`, so you can do the same thing by editing the `.profile` files.

**User Accounts Settings**
Sandboxes are _always_ restricted to the user who started them in `firejail`. You can also control _who_ can start `firejail` sandboxes with `/etc/firejail/firejail.users`.

**Comment by Raj2032** [comment]
Oh I didn't realise that you deleted your first comment as I am replying to
you via email.

Also one question, am I able to launch an application and it automatically
gets sandboxed with Firejail without having to launch the application
through firejail - so like I click on the shortcut of the application
through the DE and firejail automatically sandboxes the application?

**Comment by chiraag-nataraj** [comment]
Yes. You can use `firecfg` for that. Or, if you don't want to automatically sandbox _all_ supported applications, you can do something like `sudo ln -s /usr/bin/firejail /usr/local/bin/<program name>`.

**Comment by Raj2032** [comment]
Oh ok thanks for that :)

Does firejail reduce the program's performance, and can malicious program
break through the firejail's security and compremise the entire OS if
possible?

**Comment by chiraag-nataraj** [comment]
> Does firejail reduce the program's performance, and can malicious program break through the firejail's security and compremise the entire OS if possible?

No, firejail doesn't (generally) reduce the program's performance. Pretty much everything `firejail` uses is actually stuff implemented by the kernel (namespaces, seccomp-bpf, capabilities, etc). Nothing is perfect, so yes, something could defeat firejail's isolation techniques. But at that point, since the techniques are things enforced by the _kernel_, you have bigger issues on your hands.

Fundamentally, nothing is perfect, and security is about layers - building layers of security such that you don't depend on any one layer too much. For example, you could do firejail + apparmor + regular UNIX permissions, in which case someone would have to defeat the isolation techniques employed by firejail _and_ apparmor, as well as somehow find a way to gain root. The more you add, the harder it becomes for someone to break through, realistically speaking.

**Comment by Raj2032** [comment]
Oh I see, thanks for your help then.

**Comment by chiraag-nataraj** [comment]
No problem!

**Comment by Raj2032** [comment]
Are you the owner of Firejail?

**Comment by chiraag-nataraj** [comment]
No. I've contributed a bit of code, but @netblue30 "owns" firejail, I suppose.

**Comment by Raj2032** [comment]
Oh I see, is it written in C++?

**Comment by chiraag-nataraj** [comment]
It's written in C.

**Comment by Raj2032** [comment]
Just asking out of curosity why did you guys decide to use C over C++?
Cause I heard that with C++ you can use C features and do everything
manually without using the automated features??

**Comment by chiraag-nataraj** [comment]
I didn't make that decision. These questions are better answered by @netblue30.

**Comment by Raj2032** [comment]
Thanks for that, does he normally respond if you guys tag him?

**Comment by chiraag-nataraj** [comment]
Depending on the issue. I think this issue has strayed far from where it started off, and these questions should be resolved either by reading the FAQ and other material on the website (https://firejail.wordpress.com) or, if that doesn't answer your question, contacting netblue directly.
